### PR TITLE
Various bug fixes, early sept 2022

### PR DESCRIPTION
**Pull request by harshad1**
In this PR I fix several issues

- [x] Zim autoindent
- [x] App start tab
- [x] Text actions not being set on format change
- [x] Cleanups to patterns etc - removed redundant files and moved the hl patterns into the hl classes
- [x] Removed onViewStateRestored. Back to one function. Hopefully fewer crashes

### CODE REVIEW

**Comment by harshad1** [comment]
@gsantner There are several issues in the latest master:

1. Font is not setting correctly. It always uses the default font not matter what we select.
2. Experimental settings are not being shown no matter what the toggle is set to

**Comment by gsantner** [comment]
do we really need onViewStateRestored in documentfragment? The point is, the restored savedInstance stuff is also contained in onViewCreate. It's the same data isn't it? And onCreate&onVirwCreated are also called on rotation.

**Comment by gsantner** [comment]
>  Not sure what the source of these errors is. I will be reverting to an earlier version until they are fixed.

I saw some preferences-ktx hint in build.gradle to upgrade. Not sure if it fixes something as well.

**Comment by harshad1** [comment]
I have a feeling it's related to the context utils changes in AppSettings

Sep. 1, 2022 01:33:16 Gregor Santner ***@***.***>:

**Comment by gsantner** [comment]
> I have a feeling it's related to the context utils changes in AppSettings

Hmm that shouldn't be the case. But I check it as well when I have time.

**Comment by harshad1** [comment]
As an experiment, I reverted the AppSettings init system to just using the constructors as before and the settings issues appear fixed. I think the issue is related to the fact that `SHARED_PREFF_APP` is not used after your changes.

These reverts are in this PR

File: app/src/flavorAtest/java/net/gsantner/opoc/util/GsAndroidSupportMeWrapper.java [hunk 1]
```diff
@@ -56,11 +56,11 @@ public void mainOnResume(final Context context) {
         }
```
Comment (harshad1) [inline app/src/flavorAtest/java/net/gsantner/opoc/util/GsAndroidSupportMeWrapper.java:56]: Studio thinks this entire class / file is unused. Not sure why, but if so, do we want to keep it?

File: app/src/main/java/net/gsantner/markor/activity/ActionButtonSettingsActivity.java [hunk 2]
```diff
@@ -159,7 +160,7 @@ private Holder(View row) {
         }
 
         private void bindModel(final ActionButtonBase.ActionItem action, final String key, final Set<String> disabled) {
-            final Switch enabled = _row.findViewById(R.id.enabled_switch);
+            final SwitchCompat enabled = _row.findViewById(R.id.enabled_switch);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/ActionButtonSettingsActivity.java:163]: Studio suggested this change

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java [hunk 3]
```diff
@@ -623,13 +625,14 @@ public void applyTextFormat(final int textFormatId) {
...
-        _textActionsBar.removeAllViews();
         _textFormat = FormatRegistry.getFormat(textFormatId, activity, _document);
         _hlEditor.setHighlighter(_textFormat.getHighlighter());
         _hlEditor.setDynamicHighlightingEnabled(_appSettings.isDynamicHighlightingEnabled());
         _hlEditor.setAutoFormatters(_textFormat.getAutoFormatInputFilter(), _textFormat.getAutoFormatTextWatcher());
         _hlEditor.setAutoFormatEnabled(_appSettings.getDocumentAutoFormatEnabled(_document.getPath()));
-        _textFormat.getTextActions().setUiReferences(activity, _hlEditor, _webView);
+        _textFormat.getTextActions()
+                .setUiReferences(activity, _hlEditor, _webView)
+                .recreateTextActionBarButtons(_textActionsBar, _isPreviewVisible ? ActionButtonBase.ActionItem.DisplayMode.VIEW : ActionButtonBase.ActionItem.DisplayMode.EDIT);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java:635]: Use recreate which calls removeAllViews

File: app/src/flavorAtest/java/net/gsantner/opoc/util/GsAndroidSupportMeWrapper.java (hunk 1 above)
Comment (gsantner) [reply app/src/flavorAtest/java/net/gsantner/opoc/util/GsAndroidSupportMeWrapper.java:56]: It is currently unused yes. But you can just leave it

**Comment by gsantner** [comment]
Ok guess its just a problem then with what preference file is used by default. Shouldn't be too hard to fix :D.

**Comment by gsantner** [comment]
Ok guess its just a problem then with what preference file is used by default. Anf if it matches with the one from preferencefragment. Shouldn't be too hard to fix :D.

**Comment by harshad1** [comment]
Was there a specific reason we switched from constructors to `init`?

**Comment by gsantner** [comment]
Yesish. I wanted to make constructor private and singleton. But have not done that for the reason when really need another instance we have the option still a other option around. 

Generally said, it should not make any difference. This AppSettings creation and the AppliacationObject is the first of all that gets executed in the application overall.

File: app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java [hunk 4]
```diff
@@ -71,17 +71,14 @@ public class GsSharedPreferencesPropertyBackend implements GsPropertyBackend<Str
     protected String _prefAppName;
     protected Context _context;
 
-    public GsSharedPreferencesPropertyBackend() {
+    public GsSharedPreferencesPropertyBackend(final Context context) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java:74]: I understand that you prefer AppSettings to instantiate with Context directly. But this here can be left with non-Context requiring constructor, doesn't it?

Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java:74]: I understand that you prefer AppSettings to instantiate with Context directly. But this here can be left with non-Context requiring constructor, doesn't it? You can also call `super.init(context, name)` in AppSettings constructor.

This here so I can decide at any time to drop this way of using AppSettings and not supply context right away at instantiation. But still keep it as a single instance put somewhere. And...then initalize it there with the context later.

Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java:74]: Got it. I will rework this back, hopefully keeping the fix intact

Comment (gsantner) [reply app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java:74]: OK thanks, hope you are okay with that. But times can change fast as you know. Wouldn't be suprised if we sooner or later with Android update we cannot trust the context being available at this stage anymore. Such big breaking things unfortunately happen, just like I never thought 5 years ago that Android file system would get crippled in such a way that it is now.

Comment (gsantner) [reply app/src/main/java/net/gsantner/opoc/model/GsSharedPreferencesPropertyBackend.java:74]: > hopefully keeping the fix intact

generally said, it shouldn't make any difference at all in this case, for execution it doesn't matter if it's codeline, return, call init func, codeline, codeline,codeline...... or codeline,codeline,codeline,codeline,return

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java [hunk 5]
```diff
@@ -517,10 +517,12 @@ public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
             case R.string.action_format_todotxt:
             case R.string.action_format_plaintext:
             case R.string.action_format_markdown: {
-                _document.setFormat(itemId);
-                applyTextFormat(itemId);
-                _appSettings.setDocumentFormat(_document.getPath(), _document.getFormat());
-                return true;
+                if (itemId != _document.getFormat()) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java:520]: Only change format is we are changing

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownReplacePatternGenerator.java [hunk 6]
```diff
@@ -30,7 +31,13 @@ public class MarkdownReplacePatternGenerator {
     public static final Pattern PREFIX_UNORDERED_LIST = Pattern.compile("^(\\s*)((-|\\*|\\+)\\s)");
     public static final Pattern PREFIX_LEADING_SPACE = Pattern.compile("^(\\s*)");
 
-    static final Pattern[] PREFIX_PATTERNS = {
+    public static final AutoTextFormatter.FormatPatterns formatPatterns = new AutoTextFormatter.FormatPatterns(
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownReplacePatternGenerator.java:34]: Dramatically simplify the number of files / classes / indirection used for the patterns. We can simplify them further in the future, but this is a good stopping point.

**Comment by harshad1** [comment]
@gsantner I am done here I think - will continue testing and make a follow up if I find yet more bugs.

**Comment by harshad1** [comment]
There is 1 unit test which needs fixing. @gsantner Could you try to fix that please? If not possible, could you comment it out or something - I have will have very limited time to work on markor for the next few weeks and want to spend that on stability + bugfixes.

**Comment by gsantner** [comment]
Thank you for any time you help, doesn't matter how much, any contribution apprciated.

Don't worry about that one unit test.

**Comment by gsantner** [comment]
Thank you for any time you help, doesn't matter how much, any contribution apprciated.

Don't worry about that one unit test.

Unless this PR breaks roation (because removing onViewStateRestored)...and user can confirm no crash...I merged this soon. But feel free for new PR.  This here has important fixes.

**Comment by harshad1** [comment]
I haven't seen an issue with rotation yet. We switched to returning null within hleditor which seems to have solved the issue.

I think it's good to merge.

Sep. 2, 2022 09:34:26 Gregor Santner ***@***.***>:

File: app/src/main/java/net/gsantner/markor/activity/ActionButtonSettingsActivity.java [hunk 7]
```diff
@@ -159,7 +159,7 @@ private Holder(View row) {
...
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/ActionButtonSettingsActivity.java:162]: grafik

**Review by gsantner — CHANGES_REQUESTED** [review]

**Review by gsantner — APPROVED** [review]

**Comment by harshad1** [comment]
@gsantner do you plan to make another release soon?

**Comment by gsantner** [comment]
yes with #1835 also in

**Pull request merged**